src/pybind/mgr/cephadm/schedule.py

```diff
@@ -0,0 +1,183 @@
+
+        if self.spec.placement.host_pattern:
+            pattern_hostnames = self.spec.placement.pattern_matches_hosts(self.get_hosts_func(None))
+            if not pattern_hostnames:
+                raise OrchestratorValidationError(
+                    f'Cannot place {self.spec.one_line_str()}: No matching hosts')
+
+        if self.spec.placement.label:
+            label_hostnames = self.get_hosts_func(self.spec.placement.label)
+            if not label_hostnames:
+                raise OrchestratorValidationError(
+                    f'Cannot place {self.spec.one_line_str()}: No matching '
+                    f'hosts for label {self.spec.placement.label}')
+
+    def place(self):
+        # type: () -> List[HostPlacementSpec]
+        """
+        Load hosts into the spec.placement.hosts container.
+        """
+
+        self.validate()
+
+        count = self.spec.placement.count
+        assert count != 0
+
+        hosts = self.get_candidates()
+        if count is None:
+            logger.debug('Provided hosts: %s' % self.spec.placement.hosts)
+            return hosts
+
+        chosen = hosts
+
+
+        chosen, existing = self.prefer_hosts_with_daemons(chosen)
+
+        need = count - len(existing)
+        others = difference_hostspecs(chosen, existing)
+
+        if need < 0:
+            return self.scheduler.place(existing, count)
+        else:
+            if self.filter_new_host:
+                old = others
+                others = [h for h in others if self.filter_new_host(h.hostname)]
+                logger.debug('filtered %s down to %s' % (old, hosts))
+
+            chosen = self.scheduler.place(others, need)
+            logger.debug('Combine hosts with existing daemons %s + new hosts %s' % (
+                existing, chosen))
+            return list(merge_hostspecs(existing, chosen))
+
+    def get_candidates(self) -> List[HostPlacementSpec]:
+        if self.spec.placement.hosts:
+            return self.spec.placement.hosts
+        elif self.spec.placement.label:
+            return [
+                HostPlacementSpec(x, '', '')
+                for x in self.get_hosts_func(self.spec.placement.label)
+            ]
+        elif self.spec.placement.host_pattern:
+            return [
+                HostPlacementSpec(x, '', '')
+                for x in self.spec.placement.pattern_matches_hosts(self.get_hosts_func(None))
+            ]
+        if self.spec.placement.count is None:
+            raise OrchestratorValidationError("placement spec is empty: no hosts, no label, no pattern, no count")
+        # backward compatibility: consider an empty placements to be the same pattern = *
+        return [
+            HostPlacementSpec(x, '', '')
+            for x in self.get_hosts_func(None)
+        ]
+
+    def prefer_hosts_with_daemons(self, chosen: List[HostPlacementSpec]) -> Tuple[List[HostPlacementSpec], List[HostPlacementSpec]]:
```

Comment:
Why would we prefer hosts with daemons over empty hosts?

Reply:
otherwise we'll constantly schedule daemons on different hosts all the time. This is about keeping daemons where they are. This isn't about co-locating.